model/map-jpa/src/main/resources/META-INF/roles/jpa-roles-changelog-1.xml

```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright 2022 Red Hat, Inc. and/or its affiliates
+and other contributors as indicated by the @author tags.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd
+                    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <!-- format of id of changeSet: roles-${org.keycloak.models.map.storage.jpa.Constants.SUPPORTED_VERSION_ROLE} -->
+    <changeSet author="keycloak" id="roles-1">
+
+        <createTable tableName="role">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="metadata" type="json"/>
+        </createTable>
+        <ext:addGeneratedColumn tableName="role">
+            <ext:column name="entityversion" type="INTEGER" jsonColumn="metadata" jsonProperty="entityVersion"/>
+            <ext:column name="name" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fName"/>
+        </ext:addGeneratedColumn>
+        <createIndex tableName="role" indexName="role_entityVersion">
+            <column name="entityversion"/>
+        </createIndex>
+        <createIndex tableName="role" indexName="role_name">
+            <column name="name"/>
+        </createIndex>
```

Comment:
Could you please elaborate what do you mean by _isolation level for the JPA provider_? I'm not sure if I understand correctly.

Some indexes could be marked as unique and also some unique constraints could be added. I'll investigate possible advantages or disadvantages when using those.

note: `MapRoleProvider` does not know anything about specific implementation of the store below.

Reply:
To elaborate on the isolation level: The Map storage doesn't know about the isolation level, still the results will vary depending on the isolation configured for the storage. The following assumes that all actions are performed within a single transaction.

The code in `addRealmRole` calls `getRealmRole`, and then throws an exception if the role already exists. It then later adds the role to the store (in this case the JPA store). 

Here the varying results based on the JPA store's configuration (by reading the code, not by testing, let's hope I get it right here):

Isolation level SERAILIZABLE: The database will ensure that the search result for the realm role is stable (guaranteed not to change) within the transaction. Therefore no-one can insert another role to the database with the same name and realm (the items used in the query). Therefore the `addRealmRole` can take as long as it wants to add the new role to the database and commit the change, no-one can interfere. Anyone who tries to interfere will either be blocked or will receive an exception when trying to commit.

Any other isolation level (both REPEATABLE_READ and READ_COMMITTED): After the check that the client role is not in the database at that moment, any other parallel thread can add the role with the same name to the database and commit the change. The data would then be inconsistent. REPEATABLE_READ will only ensure that rows read from the database are unchanged (no rows have been read when looking for the existing role), and READ_COMMITTED will will only ensure that any rows read previously have been committed by their previous transactions and won't be rolled back. 

One way to prevent this inconsistency in REPEATABLE_READ or READ_COMMITTED would be a database constraint or unique index. Contrary to my previous comment, the index or constraint would need to be on realm and role, not on role alone. 

On the other hand: By having this `getRealmRole` in `addRealmRole`, the map storage provider is making an implicit assumption on the storage provider that no-one else will add a role between the check and the insert. Such implicit assumptions make it hard for a provider to do the right thing here. @hmlnarik

Reply:
I talked to Hynek yesterday about this, and the differences we will have across different providers. 

The current code in MapRoleProvider does a best-effort to check if the role exists.

A provider implementation should see what additional checks it can implement to prevent inconsistent data. For the JPA provider this would be adding constraints and unique keys. 

As this would need to be added for both clients and roles, I would be fine to add constraints in a separate issue and pull request after this PR has been merged.

Reply:
As agreed with @vramik yesterday I created a new issue for adding constraints to clients: https://github.com/keycloak/keycloak/issues/9610

Reply:
Thank you @ahus1